_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+
+The Firecracker side of this queue implementation sits right at the intersection between guest and host. According to Firecracker’s [threat model](https://github.com/firecracker-microvm/firecracker/blob/main/docs/design.md#threat-containment):
+
+
+>From a security perspective, all vCPU threads are considered to be running malicious code as soon as they have been started; these malicious threads need to be contained.
+
+
+The entirety of the VirtIO queue lives in shared memory and can thus be written to by the vCPU threads. Therefore, Firecracker cannot make any assumptions about its contents. In particular, it needs to operate securely no matter the memory content. For anyone who has worked with Kani before, this yearns for a generous application of `kani::any()`. We can set up an area of non-deterministic guest memory as follows:
+
+```rs
+fn arbitrary_guest_memory() -> GuestMemoryMmap {
+    // We need ManuallyDrop to "leak" the memory area to ensure it lives for 
+    // the entire duration of the proof
+     let memory = ManuallyDrop::new(Kani::vec::exact_vec::<u8, GUEST_MEMORY_SIZE>())
+                    .as_mut_ptr();
+    
+    let region = unsafe {
+        MmapRegionBuilder::new(GUEST_MEMORY_SIZE)
+            .with_raw_mmap_pointer(memory)
+            .build()
+            .unwrap()
+    };
+
+    let guest_region = GuestRegionMmap::new(region, GuestAddress(0)).unwrap();
+
+    // Use a single memory region, just as firecracker does for guests of size < 2GB
+    // For largest guests, firecracker uses two regions (due to the MMIO gap being
+    // at the top of 32-bit address space)
+    GuestMemoryMmap::from_regions(vec![guest_region]).unwrap()
+}
+```
+
+Note that this requires a stub for `libc::sysconf`, which is used by `.build()` to verify that guest memory is correctly aligned. We can use a stub that always returns 1, which is an over-approximation, as it causes `vm_memory` to consider all pointers to be correctly aligned.
```

Comment:
```suggestion
Note that this requires a stub for `libc::sysconf`, which is used by `.build()` to verify that guest memory is correctly aligned. We can use a stub that always returns 1, which causes `vm_memory` to consider all pointers to be correctly aligned.
```

Reply:
This is not an over-approximation, it's an implicit assumption of your harness. An over-approximation would means to consider any possible behavior of `sysconf`.